File: Plugins/AudioPlayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioTOCChapter.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.util.Log;

class AudioTOCChapter {

    private static final String TAG = "AudioTOCChapter";

    private Double[] versePositions = new Double[0];

    AudioTOCChapter(String json) {
        String trimmed = json.substring(1, json.length() - 1);
        String[] parts = trimmed.split(",");
        this.versePositions = new Double[parts.length];
        this.versePositions[0] = 0.0;
        for (int i=1; i<parts.length; i++) {
            try {
                this.versePositions[i] = Double.parseDouble(parts[i]);
            } catch (Exception ex) {
                this.versePositions[i] = this.versePositions[i-1];
            }
        }
        Log.d(TAG, this.versePositions.toString());
    }

    //deinit {
    //    print("***** Deinit AudioTOCChapter *****")
    //}

    boolean hasPositions() {
        return versePositions.length > 0;
    }

    int findVerseByPosition(int priorVerse, int milliseconds) {
        double seconds = milliseconds / 1000.0;
        int index = (priorVerse > 0 && priorVerse < this.versePositions.length) ? priorVerse : 1;
        double priorPosition = this.versePositions[index];
        if (seconds > priorPosition) {
            int lastVerse = this.versePositions.length - 1;
            while(index++ < lastVerse) {
                if (seconds < this.versePositions[index]) {
                    return (index - 1);
                }
            }
            return (lastVerse);

        } else if (seconds < priorPosition) {
            while(index-- > 2) {
                if (seconds >= this.versePositions[index]) {
                    return index;
                }
            }
            return 1;

        } else { // seconds == priorPosition
            return index;
        }
    }

    int findPositionOfVerse(int verse)  {
        double seconds = (verse > 0 && verse < this.versePositions.length) ? this.versePositions[verse] : 0.0;
        //return CMTime(seconds: seconds, preferredTimescale: CMTimeScale(1000))
        return (int)Math.floor(seconds * 1000.0);
    }

    public String toString() {
        StringBuilder str = new StringBuilder();
        for (int i = 1; i < this.versePositions.length; i++) {
            double position = this.versePositions[i];
            str.append("verse_id=");
            str.append(String.valueOf(i));
            str.append(", position=");
            str.append(String.valueOf(position));
            str.append("\n");
        }
        return(str.toString());
    }
}

File: Plugins/AudioPlayer/src/ios_oldApp/AudioPlayer/AudioTOCChapter.swift
//
//  AudioTOCChapter.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/11/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import CoreMedia

class AudioTOCChapter {
    
    private var versePositions: [Double]
    
    init(json: String) {
        let start = json.index(json.startIndex, offsetBy: 1)
        let end = json.index(json.endIndex, offsetBy: -1)
        let trimmed = json[start..<end]
        let parts = trimmed.split(separator: ",")
        self.versePositions = [Double](repeating: 0.0, count: parts.count)
        for index in 1..<parts.count {
            if let dbl = Double(parts[index]) {
                self.versePositions[index] = dbl
            }
        }
        //print(self.versePositions)
    }
    
    deinit {
        print("***** Deinit AudioTOCChapter *****")
    }
    
    func hasPositions() -> Bool {
        return versePositions.count > 0
    }
    
    func findVerseByPosition(priorVerse: Int, time: CMTime) -> Int {
         let seconds = Double(CMTimeGetSeconds(time))
        return findVerseByPosition(priorVerse: priorVerse, seconds: seconds)
    }
    
    func findVerseByPosition(priorVerse: Int, seconds: Double) -> Int {
        var index = (priorVerse > 0 && priorVerse < self.versePositions.count) ? priorVerse : 1
        let priorPosition = self.versePositions[index]
        if (seconds > priorPosition) {
            while(index < (self.versePositions.count - 1)) {
                index += 1
                let versePos = self.versePositions[index]
                if (seconds < versePos) {
                    return index - 1
                }
            }
            return self.versePositions.count - 1
            
        } else if (seconds < priorPosition) {
            while(index > 2) {
                index -= 1
                let versePos = self.versePositions[index]
                if (seconds >= versePos) {
                    return index
                }
            }
            return 1
            
        } else { // seconds == priorPosition
            return index
        }
    }
    
    func findPositionOfVerse(verse: Int) -> CMTime {
        let seconds = (verse > 0 && verse < self.versePositions.count) ? self.versePositions[verse] : 0.0
        return CMTime(seconds: seconds, preferredTimescale: CMTimeScale(1000))
    }
    
    func toString() -> String {
        var str = ""
        for (index, position) in self.versePositions.enumerated() {
            str += "verse_id=" + String(index) + ", position=" + String(position) + "\n"
        }
        return(str)
    }
}


